src/accumulator.c

```diff
@@ -0,0 +1,209 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Packetgraph.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "brick-int.h"
+#include "utils/mempool.h"
+#include "utils/bitmask.h"
+
+#define NB_POLL_MIN 5
+
+struct pg_brick *pg_accumulator_new(const char *name,
+				    enum pg_side output,
+				    struct pg_error **errp);
+
+struct pg_vec {
+	struct rte_mbuf *private_content[64];
+	int private_elem_nb;
+};
+
+static struct pg_vec pg_vecnew(void)
+{
+	struct pg_vec vec;
+
+	vec.private_elem_nb = 0;
+	return vec;
+}
+
+
+static void *pg_vecget(struct pg_vec *vec, int index)
+{
+	return vec->private_content[index];
+}
+
+static void pg_vecpush(struct pg_vec *vec, struct rte_mbuf *elem)
+{
+	struct rte_mbuf *pkt;
+
+	vec->private_content[vec->private_elem_nb] = elem;
+	pkt = pg_vecget(vec, vec->private_elem_nb);
+	pkt->udata64 = elem->udata64;
+	++vec->private_elem_nb;
+}
+
+static void pg_vecreset(struct pg_vec *vec)
+{
+	vec->private_elem_nb = 0;
+}
+
+static int pg_vecsize(struct pg_vec *vec)
+{
+	return vec->private_elem_nb;
+}
+
+struct pg_accumulator_config {
+	enum pg_side output;
+};
+
+struct pg_accumulator_state {
+	struct pg_brick brick;
+	enum pg_side output;
+	struct pg_vec burst_vec;
+	uint16_t poll_count;
+};
+
+static int accumulator_burst(struct pg_brick *brick, enum pg_side from,
+			     uint16_t edge_index, struct rte_mbuf **pkts,
+			     uint64_t pkts_mask, struct pg_error **errp)
+{
+	struct pg_accumulator_state *state;
+	struct pg_brick_side *s;
+	bool burst = false;
+	int size;
+
+	state = pg_brick_get_state(brick, struct pg_accumulator_state);
+	s = &brick->sides[pg_flip_side(from)];
+	PG_FOREACH_BIT(pkts_mask, it) {
+		size = pg_vecsize(&state->burst_vec);
+		if (size == 64 || state->poll_count == NB_POLL_MIN) {
+			pg_brick_burst(s->edge.link,
+				       from, s->edge.pair_index,
+				       state->burst_vec.private_content,
+				       pg_mask_firsts(size),
+				       errp);
+			if (state->poll_count == NB_POLL_MIN) {
+				state->poll_count = 0;
+				return 0;
+			}
+			burst = true;
+			pg_vecreset(&state->burst_vec);
+		}
+		pg_vecpush(&state->burst_vec, pkts[it]);
+	}
+	if (!burst)
+		pg_brick_burst(s->edge.link,
+			       from, s->edge.pair_index,
+			       NULL,
+			       0,
+			       errp);
+	return 0;
+}
+
+static int accumulator_poll(struct pg_brick *brick,
+			    uint16_t *pkts_cnt,
+			    struct pg_error **errp)
+{
+	struct pg_accumulator_state *state;
+	struct pg_brick_side *s;
+	uint64_t pkts_mask;
+
+	state = pg_brick_get_state(brick, struct pg_accumulator_state);
+	s = &brick->sides[state->output];
+
+	if (unlikely(s->edge.link == NULL))
+		return 0;
+
+	if (state->poll_count < NB_POLL_MIN) {
+		++state->poll_count;
+		return 0;
+	}
+	pkts_mask = pg_mask_firsts(pg_vecsize(&state->burst_vec));
```

Comment:
I'm not a big fan of calling pg_mask_firsts at each poll
a good hack can be to do
pkts_mask = (1 << pg_vecsize(&state->burst_vec)) - 1;